### PR TITLE
About me page redesign

### LINKED ISSUE #206: 🎨 Redesign the about page
I want to redesign the about me page since there are now contributors added to it 🤔 I'll post WIPs to get the feedback

Blocked by https://github.com/brendt/rfc-vote/pull/205

### PR DESCRIPTION
**Pull request by SerhiiCho**
About me page redesign. Here are screenshots:

closes #206

## Desktop
image

image

### CODE REVIEW

**Comment by SerhiiCho** [comment]
I've made contributors in random order to make so that the position in `contributors.json` does not impact the position on the page.

**Comment by Plytas** [comment]
Looks beautiful!

File: app/Http/Controllers/AboutController.php [hunk 1]
```diff
@@ -4,18 +4,35 @@
...
 {
+    private const DAY = 60 * 60 * 24;
+
     public function __invoke(): View
     {
-        $contributors = array_map(
-            fn (array $item) => new Contributor(...$item),
-            json_decode(file_get_contents(__DIR__.'/../../../contributors.json') ?: '{}', true)['contributors'] ?? [],
-        );
+        $contributors = Cache::remember('contributors', self::DAY, fn () => $this->getContributors());
```
Comment (Plytas) [inline app/Http/Controllers/AboutController.php:16]:
```suggestion
        $contributors = Cache::remember('contributors', self::DAY, $this->getContributors(...));
```

File: app/Http/Controllers/AboutController.php [hunk 2]
```diff
@@ -4,18 +4,35 @@
...
+
+    /**
+     * @return Contributor[]
+     *
+     * @throws JsonException
+     */
+    private function getContributors(): array
+    {
+        $content = file_get_contents(__DIR__.'/../../../contributors.json') ?: '{}';
+        $people = json_decode($content, true, 512, JSON_THROW_ON_ERROR);
```
Comment (Plytas) [inline app/Http/Controllers/AboutController.php:31]:
```suggestion
        /** @var array{contributors: array<int, array{id: int, name: string, url: string, contributions: array<int, string>}>} $contributors */
        $contributors = File::json(__DIR__.'/../../../contributors.json', JSON_THROW_ON_ERROR);
```

File: app/Http/Controllers/AboutController.php [hunk 3]
```diff
@@ -4,18 +4,35 @@
...
 
-        return view('about', [
-            'contributors' => $contributors,
-        ]);
+        return array_map(
+            static fn (array $item) => new Contributor(...$item),
+            $people['contributors'] ?? [],
+        );
```
Comment (Plytas) [inline app/Http/Controllers/AboutController.php:36]:
```suggestion
        return collect($contributors['contributors'])
            ->map(fn (array $contributor) => new Contributor(...$contributor))
            ->all();
```

File: contributors.json [hunk 4]
```diff
@@ -1,102 +1,127 @@
 {
     "contributors": [
         {
+            "id": 6905297,
```
Comment (Plytas) [inline contributors.json:4]: Where do these IDs come from?

**Review by Plytas — COMMENTED** [review]
Suggesting some refactoring for newer cleaner syntax.

**Comment by MohammadAlhallaq** [comment]
this is beautifull! 🤩

**Comment by vsergiu93** [comment]
Looks awesome

File: contributors.json (hunk 4 above)
Comment (SerhiiCho) [reply contributors.json:4]: Github, I manually added for everyone

Comment (Plytas) [reply contributors.json:4]: Got it, smart!

File: app/Http/Controllers/AboutController.php (hunk 3 above)
Comment (SerhiiCho) [reply app/Http/Controllers/AboutController.php]: Wow! Much better

File: app/Http/Controllers/AboutController.php (hunk 1 above)
Comment (Plytas) [reply app/Http/Controllers/AboutController.php]: Since PHP 8.1 https://stitcher.io/blog/new-in-php-81#first-class-callable-syntax-rfc

**Comment by Plytas** [comment]
These phpstan errors will be fixed if you merge the last suggestion.

File: app/Http/Controllers/AboutController.php (hunk 2 above)
Comment (SerhiiCho) [reply app/Http/Controllers/AboutController.php]: Great catch, didn't notice that...

**Comment by SerhiiCho** [comment]
Thank you @Plytas for the feedback, great suggestions

**Comment by SerhiiCho** [comment]
I don't know how I feel about the GitHub bot fixing styles.. It pushes changes to my working branch and it conflicts with my local work. Am I the only one who doesn't like it?

**Comment by Plytas** [comment]
Definitely get in the habit of running `composer qa` or just `composer pint` before committing. We could even consider adding pint as pre-commit hook. Apart from that without the bot there would be no way to maintain consistent styling, because someone will forget to run it locally.

**Comment by SerhiiCho** [comment]
> there would be no way to maintain consistent styling

 I think there is a way to tell Pint so that instead of fixing styles to just error out. It will fail the checks on PR. Anyway, in which PHP version we got this syntax with 3 dots?:

```php
Cache::remember('contributors', self::DAY, $this->getContributors(...))
```
I somehow missed it

**Comment by Plytas** [comment]
>in which PHP version we got this syntax with 3 dots?

I commented this on the suggestion, but it might have gotten lost. We have it since PHP 8.1 https://stitcher.io/blog/new-in-php-81#first-class-callable-syntax-rfc

**Comment by Plytas** [comment]
>I think there is a way to tell Pint so that instead of fixing styles to just error out. It will fail the checks on PR.

You are right about that, we could call the action with `--test` flag and then it would only error out. I feel like that would be more work for @brendt in case original PR author disappears and doesn't push pint fixes. He would have to checkout the PR, run pint and push it himself.

File: app/Http/Controllers/AboutController.php [hunk 5]
```diff
@@ -4,18 +4,44 @@
...
     {
-        $contributors = array_map(
-            fn (array $item) => new Contributor(...$item),
-            json_decode(file_get_contents(__DIR__.'/../../../contributors.json') ?: '{}', true)['contributors'] ?? [],
-        );
-
-        return view('about', [
-            'contributors' => $contributors,
-        ]);
+        $contributors = Cache::remember('contributors', self::DAY, $this->getContributors(...));
```
Comment (joshbonnick) [inline app/Http/Controllers/AboutController.php:17]:
```suggestion
        $contributors = Cache::remember('contributors', now()->addDay(), $this->getContributors(...));
```

File: app/Http/Controllers/AboutController.php [hunk 6]
```diff
@@ -4,18 +4,44 @@
...
+         * @var array{
+         *     contributors: array<int, array{
+         *         id: int,
+         *         name: string,
+         *         url: string,
+         *         contributions: array<int, string>,
+         *     }>
+         * } $contributors
+         */
+        $contributors = File::json(__DIR__.'/../../../contributors.json', JSON_THROW_ON_ERROR);
```
Comment (joshbonnick) [inline app/Http/Controllers/AboutController.php:41]:
```suggestion
        $contributors = File::json(base_path('contributors.json'), JSON_THROW_ON_ERROR);
```

Comment (Plytas) [reply app/Http/Controllers/AboutController.php:41]: I forgot about `base_path()`, was looking for `root_path()` but couldn't find it 🤦

**Comment by joshbonnick** [comment]
It may also be worth extracting `getContributers` to another class, perhaps `FetchContributers` in the Support directory. We currently cannot mock the list of contributors, nor retrieve it for output elsewhere.

**Comment by brendt** [comment]
> You are right about that, we could call the action with --test flag and then it would only error out. I feel like that would be more work for @brendt in case original PR author disappears and doesn't push pint fixes. He would have to checkout the PR, run pint and push it himself.

I'm ok with that, I agree that dealing with style commits can be a hassle. Made an issue for it #239, I'll probably fix it tomorrow.

On the other hand: always best to run `composer ca` before committing. 

> It may also be worth extracting getContributers to another class, perhaps FetchContributers in the Support directory. We currently cannot mock the list of contributors, nor retrieve it for output elsewhere.

@joshbonnick good idea, could you make an issue for it so that we don't forget?

@SerhiiCho the page looks 🔥 can I merge this, or do you need to make changes still?

**Comment by SerhiiCho** [comment]
You can merge it

**Comment by SerhiiCho** [comment]
> It may also be worth extracting getContributers to another class

Great idea, but in a separate PR

**Comment by brendt** [comment]
It's live :) I did notice Nikita's avatar not showing up, probably an uncomitted file?

**Comment by Plytas** [comment]
`artisan storage:link` maybe?

**Comment by joshbonnick** [comment]
> It's live :) I did notice Nikita's avatar not showing up, probably an uncommitted file?

It's in `.gitignore`

**Comment by Plytas** [comment]
I'm pretty sure it's just storage link, as the file is definitely committed, but `asset()` is looking in public dir.

**Comment by SerhiiCho** [comment]
I'm also thinking that you just need to run `php artisan storage:link`

**Comment by brendt** [comment]
Ah, the deploy script we're using clears everything from `/public/public`, and symlinks it to the storage folder on the server for persistent data. In other words: you cannot use that folder 😁 I've simply moved it to `/public` for now, and it works just fine :)

**Pull request merged**